### PR TITLE
feat(go): support self hosted go modules in gitlab

### LINKED ISSUE #8859: Support self-hosted Go Modules in GitLab
**What Renovate type, platform and version are you using?**

- Self-hosted Renovate
- Version: `renovate/renovate:24.63.0@sha256:b79ce7d09a53f0f4b1ffe7ec1e91a0c8f088e1858d045c83dcf8eeba5a306a46` 
- Gitlab EE 

**Describe the bug**

Renovate is unable to lookup dependencies that exist in our private, self hosted Gitlab EE repositories. When an onboarding MR is raised we have warnings only for our own go mod dependencies, everything else (docker, github, etc) is fine.

**Relevant debug logs**

Appropriate log from renovate (i've removed domain name): 

```
DEBUG: Failed to look up dependency [domain_name]/go/driver (repository=go/driver, packageFile=go.mod, dependency=[domain]/go/driver)
DEBUG: Go lookup source url (repository=go/driver)
       "goModule": "[domain]/go/go-curve/v2",
       "goSourceUrl": "https://[domain]/go/go-curve"
DEBUG: Datasource unknown error (repository=go/driver)
       "datasource": "go",
       "lookupName": "[domain]/go/go-curve/v2",
       "err": {
         "message": "Cannot read property 'datasource' of null",
         "stack": "TypeError: Cannot read property 'datasource' of null\n    at Object.getReleases (/usr/src/app/node_modules/renovate/lib/datasource/go/index.ts:128:18)\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)\n    at fetchReleases (/usr/src/app/node_modules/renovate/lib/datasource/index.ts:218:13)\n    at Object.getPkgReleases (/usr/src/app/node_modules/renovate/lib/datasource/index.ts:272:7)\n    at Object.lookupUpdates (/usr/src/app/node_modules/renovate/lib/workers/repository/process/lookup/index.ts:55:30)\n    at fetchDepUpdates (/usr/src/app/node_modules/renovate/lib/workers/repository/process/fetch.ts:42:26)\n    at /usr/src/app/node_modules/p-map/index.js:57:22"
       }
```

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/master/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] This is a really small bug, it does not need a reproduction (think small typo)
- [ ] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [x] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

Looking at the logs, I could see a particularly interesting line `DEBUG: Go lookup source url`. I found that message in the source here, https://github.com/renovatebot/renovate/blob/6b15f6ef68c18336562aa1490d7216e304fd91d9/lib/datasource/go/index.ts#L61 and determined the following. 

The `if (sourceMatch) {` is true as when calling `https://${goModule}?go-get=1` you do get back the go-import and go-source meta html. 

What seems to happen though is that the following two if statements are not satisfied as the URL is not a github.com URL, and nor is it `/^(https:\/\/[^/]*gitlab.[^/]*)\/(.*)$/`. This causes `getDatasource` to return null. 

```
f (goSourceUrl?.startsWith('https://github.com/')) {
      return {
        datasource: github.id,
        lookupName: goSourceUrl
          .replace('https://github.com/', '')
          .replace(/\/$/, ''),
      };
    }
    const gitlabRes = gitlabRegExp.exec(goSourceUrl);
    if (gitlabRes) {
      return {
        datasource: gitlab.id,
        registryUrl: gitlabRes[1],
        lookupName: gitlabRes[2].replace(/\/$/, ''),
      };
    }
```

Perhaps a resolution here could be to default to the platform that is specified in config, in our case `gitlab`? Or use the `RENOVATE_ENDPOINT` to configure the `gitlabRegExp` for a best guess attempt at resolving the dependency? Obviously once that is working the call to Gitlab needs to be made with the PAT to be able to read from the private repository. 

So far i'm really impressed with Renovate and the flexibility it provides, we have hundreds of repositories we want to automatically manage our dependencies in but the main use case we have is allowing us to manage internal dependencies updates which it is currently failing to do. Any guidance / suggestions would be really appreciated :)

### PR DESCRIPTION
**Pull request by davidgwcurve**
## Changes:

Fixes issue #8859, https://github.com/renovatebot/renovate/issues/8859. 

```
- Query hostRules for hostType=gitlab, url=https://[go module]
- If the answer back from hostRules contains a token field, then we know it's a GitLab host and can redirect the lookup to a gitlab tags one
```

Also fixes the same logic in the changelog functionality, as after this issue was resolved I got 404's. 

## Context:

Closes #8859

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added unit tests, or **(TODO)**
- [ ] No new tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by HonkingGoose** [comment]
@davidgwcurve Your latest commit includes changes to `package.json` and `yarn.lock` files, was that intended?

**Review by rarkins — CHANGES_REQUESTED** [review]
Revert package json changes

File: lib/datasource/go/index.spec.ts [hunk 1]
```diff
@@ -216,19 +254,19 @@ describe('datasource/go', () => {
       httpMock
         .scope('https://some.unknown.website/')
         .get('/example/module?go-get=1')
-        .reply(404);
+        .reply(200);
```
Comment (davidgwcurve) [inline lib/datasource/go/index.spec.ts:257]: there is a test that checks for a 404 returned from the specified URL already, this test should actually be checking that the call was successful but the code doesn't know how to handle the dependency (i.e. not github/gitlab.com/not a gitlab hostRules)

**Comment by davidgwcurve** [comment]
@rarkins I've added unit tests now and ran on our own repository, it works as expected.

**Review by rarkins — APPROVED** [review]

**Pull request merged**